hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/InsertIntoHoodieTableCommand.scala

```diff
@@ -0,0 +1,247 @@
+    val dataProjects = queryDataFields.zip(targetDataSchema.fields).map {
+      case (dataAttr, targetField) =>
+        val castAttr = castIfNeeded(dataAttr,
+          targetField.dataType, conf)
+        new Column(Alias(castAttr, targetField.name)())
+    }
+
+    val partitionProjects = if (staticPartitionValues.isEmpty) { // insert dynamic partitions
+      // The partition attributes is followed the data attributes in the query
+      // So we init the partitionAttrPosition with the data schema size.
+      var partitionAttrPosition = targetDataSchema.size
+      targetPartitionSchema.fields.map(f => {
+        val partitionAttr = query.logicalPlan.output(partitionAttrPosition)
+        partitionAttrPosition = partitionAttrPosition + 1
+        // Cast the partition attribute to the target table field's type.
+        val castAttr = castIfNeeded(partitionAttr, f.dataType, conf)
+        new Column(Alias(castAttr, f.name)())
+      })
+    } else { // insert static partitions
+      targetPartitionSchema.fields.map(f => {
+        val staticPartitionValue = staticPartitionValues.getOrElse(f.name,
+        s"Missing static partition value for: ${f.name}")
+        // Cast the static partition value to the target table field's type.
+        val castAttr = castIfNeeded(
+          Literal.create(staticPartitionValue, StringType), f.dataType, conf)
+        new Column(Alias(castAttr, f.name)())
+      })
+    }
+    val alignedProjects = dataProjects ++ partitionProjects
+    query.select(alignedProjects: _*)
+  }
+
+  /**
+    * Build the default config for insert.
+    * @return
+    */
+  private def buildHoodieInsertConfig(table: CatalogTable,
+                              sparkSession: SparkSession,
+                              insertPartitions: Map[String, Option[String]] = Map.empty,
+                              schema: Option[StructType] = None): Map[String, String] = {
+
+    if (insertPartitions.nonEmpty &&
+      (insertPartitions.keys.toSet != table.partitionColumnNames.toSet)) {
+      throw new IllegalArgumentException(s"Insert partition fields" +
+        s"[${insertPartitions.keys.mkString(" " )}]" +
+        s" not equal to the defined partition in table[${table.partitionColumnNames.mkString(",")}]")
+    }
+    val parameters = HoodieOptionConfig.mappingSqlOptionToHoodieParam(table.storage.properties)
+
+    val tableType = parameters.getOrElse(TABLE_TYPE_OPT_KEY, DEFAULT_TABLE_TYPE_OPT_VAL)
+
+    val partitionFields = table.partitionColumnNames.mkString(",")
+    val path = getTableLocation(table, sparkSession)
+      .getOrElse(s"Missing location for table ${table.identifier}")
+
+    val tableSchema = schema.getOrElse(table.schema)
+    val options = table.storage.properties
+    val primaryColumns = HoodieOptionConfig.getPrimaryColumns(options)
+
+    val keyGenClass = if (primaryColumns.nonEmpty) {
+      classOf[ComplexKeyGenerator].getCanonicalName
+    } else {
+      classOf[UuidKeyGenerator].getName
```

Comment:
> I noticed current implementation allows creating a table without a primary key and allow to insert but not update or delete, I wonder what's use case of this?

For the append-only case, which user just want to insert data to the table without pk constraints，we need allow to define the non-pk table. e.g. user write the log data to the hudi, in the case ,the log data has no pk at all.

> I couldn't find the way to alter table with storage.properties, I wanted to try changing non-primary-key table into primary-key table, is this possible?

Yes, since this is the first PR for the sql support. The Alter command for hudi is not included in the PR. I will submit another PR to support this.
I think we should not allow changing from non-primary-key table to a primary-key table. Because it will change the table structure.